File: scripts/publish-skills.mjs
#!/usr/bin/env node

/**
 * Publish all skills from a directory to the iKhono registry.
 *
 * Scans a folder for subdirectories containing skill.yaml + SKILL.md,
 * then publishes each one via the iKhono API.
 *
 * Usage:
 *   node scripts/publish-skills.mjs                          # publishes everything in examples/
 *   node scripts/publish-skills.mjs --dir ./my-skills        # custom skills directory
 *   node scripts/publish-skills.mjs --changelog "Bug fixes"  # set changelog for all
 *   node scripts/publish-skills.mjs --dry-run                # preview without publishing
 *
 * Auth:
 *   Reads token from ~/.ikhono/config.json (set via `ikh login`).
 *   Override with --token <token> or IKH_TOKEN env var.
 *
 * API:
 *   Defaults to https://ikhono.io. Override with --api-url or IKH_API_URL env var.
 */

import { readdir, readFile } from 'node:fs/promises';
import { join, resolve } from 'node:path';
import { homedir } from 'node:os';
import { existsSync } from 'node:fs';

// --- CLI args ---

const args = process.argv.slice(2);

function getArg(name) {
  const idx = args.indexOf(`--${name}`);
  if (idx === -1 || idx + 1 >= args.length) return undefined;
  return args[idx + 1];
}

function hasFlag(name) {
  return args.includes(`--${name}`);
}

if (hasFlag('help') || hasFlag('h')) {
  console.log(`Usage: node scripts/publish-skills.mjs [options]

Options:
  --dir <path>        Skills directory (default: examples/ next to this script)
  --changelog <text>  Changelog message for all skills
  --token <token>     API token (default: from ~/.ikhono/config.json)
  --api-url <url>     API base URL (default: from config or https://ikhono.io)
  --dry-run           Preview what would be published without making API calls
  --help              Show this help message
`);
  process.exit(0);
}

const DRY_RUN = hasFlag('dry-run');
const CHANGELOG = getArg('changelog') || '';
const SKILLS_DIR = resolve(getArg('dir') || join(import.meta.dirname, '..', 'examples'));

// --- Config ---

function loadConfig() {
  const configPath = join(homedir(), '.ikhono', 'config.json');
  if (!existsSync(configPath)) return {};
  try {
    const raw = JSON.parse(readFileSync(configPath, 'utf-8'));
    return raw;
  } catch {
    return {};
  }
}

// Need sync readFile for config — import separately
import { readFileSync } from 'node:fs';

const config = loadConfig();
const TOKEN = getArg('token') || process.env.IKH_TOKEN || config.token;
const API_URL = getArg('api-url') || process.env.IKH_API_URL || config.apiUrl || 'https://ikhono.io';

if (!TOKEN && !DRY_RUN) {
  console.error('No auth token found. Run `ikh login` first, or pass --token <token>');
  process.exit(1);
}

// --- YAML parser (same as CLI) ---

function parseSimpleYaml(yaml) {
  const result = {};
  let currentKey = '';
  let currentArray = null;

  for (const line of yaml.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;

    if (trimmed.startsWith('- ') && currentArray !== null) {
      currentArray.push(trimmed.slice(2).replace(/^["']|["']$/g, ''));
      result[currentKey] = currentArray;
      continue;
    }

    if (currentArray !== null) {
      currentArray = null;
    }

    const match = trimmed.match(/^(\w+):\s*(.*)?$/);
    if (match) {
      const [, key, value] = match;
      if (!value || value === '') {
        currentKey = key;
        currentArray = [];
      } else {
        result[key] = value.replace(/^["']|["']$/g, '');
      }
    }
  }

  return result;
}

// --- Publish ---

async function publishSkill(skillDir, name) {
  const yamlPath = join(skillDir, 'skill.yaml');
  const mdPath = join(skillDir, 'SKILL.md');

  let yamlText, content;
  try {
    yamlText = await readFile(yamlPath, 'utf-8');
    content = await readFile(mdPath, 'utf-8');
  } catch {
    console.log(`  skip: ${name}/ — missing skill.yaml or SKILL.md`);
    return { status: 'skipped', name };
  }

  const parsed = parseSimpleYaml(yamlText);

  if (!parsed.name || !parsed.description) {
    console.log(`  skip: ${name}/ — skill.yaml missing name or description`);
    return { status: 'skipped', name };
  }

  if (DRY_RUN) {
    console.log(`  would publish: ${parsed.name} v${parsed.version || '1.0.0'} — ${parsed.description}`);
    return { status: 'dry-run', name: parsed.name };
  }

  const body = {
    skillYaml: JSON.stringify(parsed),
    content: content.trim(),
    ...(CHANGELOG && { changelog: CHANGELOG }),
  };

  try {
    const res = await fetch(`${API_URL}/api/skills`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${TOKEN}`,
      },
      body: JSON.stringify(body),
    });

    if (!res.ok) {
      const err = await res.json().catch(() => ({ error: res.statusText }));
      if (res.status === 409) {
        console.log(`  skip: ${parsed.name} v${parsed.version || '1.0.0'} — version already exists`);
        return { status: 'skipped', name: parsed.name };
      }
      console.log(`  fail: ${parsed.name} — ${JSON.stringify(err.error)}`);
      return { status: 'failed', name: parsed.name, error: err.error };
    }

    const { data } = await res.json();
    console.log(`  published: ${data.slug} v${data.version} — ${data.message}`);
    return { status: 'published', name: data.slug, version: data.version };
  } catch (err) {
    console.log(`  fail: ${parsed.name} — ${err.message}`);
    return { status: 'failed', name: parsed.name, error: err.message };
  }
}

async function main() {
  console.log(`Skills dir: ${SKILLS_DIR}`);
  console.log(`API:        ${API_URL}`);
  console.log(`Mode:       ${DRY_RUN ? 'dry-run' : 'publish'}\n`);

  let entries;
  try {
    entries = await readdir(SKILLS_DIR, { withFileTypes: true });
  } catch {
    console.error(`Cannot read directory: ${SKILLS_DIR}`);
    process.exit(1);
  }

  const dirs = entries.filter((e) => e.isDirectory()).sort((a, b) => a.name.localeCompare(b.name));

  const results = [];

  if (dirs.length === 0) {
    // No subdirectories — check if the dir itself is a skill
    const hasYaml = entries.some((e) => e.name === 'skill.yaml');
    const hasMd = entries.some((e) => e.name === 'SKILL.md');
    if (hasYaml && hasMd) {
      console.log('Found single skill in directory\n');
      results.push(await publishSkill(SKILLS_DIR, SKILLS_DIR.split('/').pop()));
    } else {
      console.log('No skill directories found and no skill.yaml + SKILL.md in this directory.');
      process.exit(0);
    }
  } else {
    console.log(`Found ${dirs.length} skill(s)\n`);
    for (const entry of dirs) {
      results.push(await publishSkill(join(SKILLS_DIR, entry.name), entry.name));
    }
  }

  // Summary
  const published = results.filter((r) => r.status === 'published').length;
  const failed = results.filter((r) => r.status === 'failed').length;
  const skipped = results.filter((r) => r.status === 'skipped').length;
  const dryRun = results.filter((r) => r.status === 'dry-run').length;

  console.log(`\nDone: ${published} published, ${failed} failed, ${skipped} skipped${dryRun ? `, ${dryRun} dry-run` : ''}`);

  if (failed > 0) process.exit(1);
}

main();

File: scripts/publish-skills.py
#!/usr/bin/env python3

"""
Publish all skills from a directory to the iKhono registry.

Scans a folder for subdirectories containing skill.yaml + SKILL.md,
then publishes each one via the iKhono API.

Usage:
  python scripts/publish-skills.py                          # publishes everything in examples/
  python scripts/publish-skills.py --dir ./my-skills        # custom skills directory
  python scripts/publish-skills.py --changelog "Bug fixes"  # set changelog for all
  python scripts/publish-skills.py --dry-run                # preview without publishing

Auth:
  Reads token from ~/.ikhono/config.json (set via `ikh login`).
  Override with --token <token> or IKH_TOKEN env var.

API:
  Defaults to https://ikhono.io. Override with --api-url or IKH_API_URL env var.
"""

import argparse
import json
import os
import re
import sys
import urllib.request
import urllib.error
from pathlib import Path


def load_config():
    config_path = Path.home() / ".ikhono" / "config.json"
    if not config_path.exists():
        return {}
    try:
        return json.loads(config_path.read_text())
    except (json.JSONDecodeError, OSError):
        return {}


def parse_simple_yaml(text):
    """Minimal YAML parser for flat key-value pairs and string arrays."""
    result = {}
    current_key = ""
    current_array = None

    for line in text.splitlines():
        trimmed = line.strip()
        if not trimmed or trimmed.startswith("#"):
            continue

        if trimmed.startswith("- ") and current_array is not None:
            value = trimmed[2:].strip().strip("\"'")
            current_array.append(value)
            result[current_key] = current_array
            continue

        if current_array is not None:
            current_array = None

        match = re.match(r"^(\w+):\s*(.*)?$", trimmed)
        if match:
            key, value = match.group(1), match.group(2) or ""
            if not value:
                current_key = key
                current_array = []
            else:
                result[key] = value.strip("\"'")

    return result


def publish_skill(skill_dir, name, api_url, token, changelog, dry_run):
    yaml_path = skill_dir / "skill.yaml"
    md_path = skill_dir / "SKILL.md"

    if not yaml_path.exists() or not md_path.exists():
        print(f"  skip: {name}/ — missing skill.yaml or SKILL.md")
        return "skipped"

    parsed = parse_simple_yaml(yaml_path.read_text())
    content = md_path.read_text().strip()

    if not parsed.get("name") or not parsed.get("description"):
        print(f"  skip: {name}/ — skill.yaml missing name or description")
        return "skipped"

    version = parsed.get("version", "1.0.0")

    if dry_run:
        print(f"  would publish: {parsed['name']} v{version} — {parsed['description']}")
        return "dry-run"

    body = {
        "skillYaml": json.dumps(parsed),
        "content": content,
    }
    if changelog:
        body["changelog"] = changelog

    data = json.dumps(body).encode()
    req = urllib.request.Request(
        f"{api_url}/api/skills",
        data=data,
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {token}",
            "User-Agent": "ikhono-publish-skills/1.0",
        },
        method="POST",
    )

    try:
        with urllib.request.urlopen(req) as res:
            result = json.loads(res.read())["data"]
            print(f"  published: {result['slug']} v{result['version']} — {result['message']}")
            return "published"
    except urllib.error.HTTPError as e:
        if e.code == 409:
            print(f"  skip: {parsed['name']} v{version} — version already exists")
            return "skipped"
        try:
            err = json.loads(e.read())
            print(f"  fail: {parsed['name']} — {json.dumps(err.get('error', e.reason))}")
        except Exception:
            print(f"  fail: {parsed['name']} — {e.reason}")
        return "failed"
    except Exception as e:
        print(f"  fail: {parsed['name']} — {e}")
        return "failed"


def main():
    parser = argparse.ArgumentParser(description="Publish skills to the iKhono registry")
    parser.add_argument("--dir", help="Skills directory (default: examples/ next to this script)")
    parser.add_argument("--changelog", default="", help="Changelog message for all skills")
    parser.add_argument("--token", help="API token (default: from ~/.ikhono/config.json)")
    parser.add_argument("--api-url", help="API base URL (default: from config or https://ikhono.io)")
    parser.add_argument("--dry-run", action="store_true", help="Preview without publishing")
    args = parser.parse_args()

    config = load_config()
    token = args.token or os.environ.get("IKH_TOKEN") or config.get("token")
    api_url = args.api_url or os.environ.get("IKH_API_URL") or config.get("apiUrl") or "https://ikhono.io"
    api_url = api_url.rstrip("/")

    if not token and not args.dry_run:
        print("No auth token found. Run `ikh login` first, or pass --token <token>", file=sys.stderr)
        sys.exit(1)

    script_dir = Path(__file__).resolve().parent
    skills_dir = Path(args.dir).resolve() if args.dir else script_dir.parent / "examples"

    print(f"Skills dir: {skills_dir}")
    print(f"API:        {api_url}")
    print(f"Mode:       {'dry-run' if args.dry_run else 'publish'}\n")

    if not skills_dir.is_dir():
        print(f"Cannot read directory: {skills_dir}", file=sys.stderr)
        sys.exit(1)

    dirs = sorted([d for d in skills_dir.iterdir() if d.is_dir()])
    results = []

    if not dirs:
        # No subdirectories — check if the dir itself is a skill
        if (skills_dir / "skill.yaml").exists() and (skills_dir / "SKILL.md").exists():
            print("Found single skill in directory\n")
            results.append(publish_skill(skills_dir, skills_dir.name, api_url, token, args.changelog, args.dry_run))
        else:
            print("No skill directories found and no skill.yaml + SKILL.md in this directory.")
            sys.exit(0)
    else:
        print(f"Found {len(dirs)} skill(s)\n")
        for d in dirs:
            results.append(publish_skill(d, d.name, api_url, token, args.changelog, args.dry_run))

    # Summary
    published = results.count("published")
    failed = results.count("failed")
    skipped = results.count("skipped")
    dry_run = results.count("dry-run")

    summary = f"\nDone: {published} published, {failed} failed, {skipped} skipped"
    if dry_run:
        summary += f", {dry_run} dry-run"
    print(summary)

    if failed > 0:
        sys.exit(1)


if __name__ == "__main__":
    main()
